modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/snapshot/RocksSnapshotManager.java

```diff
@@ -77,18 +78,22 @@ public RocksSnapshotManager(RocksDB db, Collection<ColumnFamilyRange> ranges, Ex
     public CompletableFuture<Void> createSnapshot(Path snapshotDir) {
         Path tmpPath = Paths.get(snapshotDir.toString() + TMP_SUFFIX);
 
-        return CompletableFuture.supplyAsync(db::getSnapshot, executor)
-                .thenComposeAsync(snapshot -> {
+        // The snapshot reference must be taken synchronously, otherwise we might let more writes sneak into the snapshot than needed.
+        Snapshot snapshot = db.getSnapshot();
+
+        return CompletableFuture.supplyAsync(
+                () -> {
                     createTmpSnapshotDir(tmpPath);
 
                     // Create futures for capturing SST snapshots of the column families
                     CompletableFuture<?>[] sstFutures = ranges.stream()
                             .map(cf -> createSstFileAsync(cf, snapshot, tmpPath))
                             .toArray(CompletableFuture[]::new);
 
-                    return CompletableFuture.allOf(sstFutures).thenApply(v -> snapshot);
+                    return CompletableFuture.allOf(sstFutures);
                 }, executor)
-                .whenCompleteAsync((snapshot, e) -> {
+                .thenCompose(Function.identity())
+                .whenCompleteAsync((ignored, e) -> {
                     if (e != null) {
```

Comment:
Can you please also move this statement below the closing of the snapshot? Thanks!

Reply:
Thanks for catching this